clang/compiler-course/kurakin_user_data_type/kurakin_user_data_type_plugin.cpp

```diff
@@ -0,0 +1,86 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+class UserDataTypeVisitor final : public clang::RecursiveASTVisitor<UserDataTypeVisitor> {
+public:
+  explicit UserDataTypeVisitor(clang::ASTContext *context) : m_context(context) {}
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *rd) {
+    llvm::outs() << rd->getNameAsString();
+    for (auto it = rd->bases_begin(); it != rd->bases_end(); ++it) {
+      if(it!=rd->bases_begin()) llvm::outs() << ",";
+      if(it == rd->bases_begin()) llvm::outs() << " -> ";
+      llvm::outs() << it->getType().getAsString();
+    }
+    llvm::outs() << "\n";
+
+    llvm::outs() << "|_Fields\n";
+    for(auto f : rd->fields()){
+      llvm::outs() << "| |_ " << f->getName() << " (" << f->getType().getAsString();
+      if(f->getAccess() == clang::AS_public) llvm::outs() << "|public";
+      else if(f->getAccess() == clang::AS_private) llvm::outs() << "|private";
+      else if(f->getAccess() == clang::AS_protected) llvm::outs() << "|protected";
+      llvm::outs() << ")\n";
+    }
+
+    llvm::outs() << "|\n";
+
+    llvm::outs() << "|_Methods\n";
+    for(auto m : rd->methods()){
+      if(m->isImplicit()) continue;
+
+      llvm::outs() << "| |_ " << m->getNameAsString() << " (" << m->getReturnType().getAsString() << "(";
+      for(size_t i=0; i < m->getNumParams(); ++i){
+        if(i!=0) llvm::outs() << ",";
+        llvm::outs() << m->parameters()[i]->getType().getAsString();
+      }
+      llvm::outs() << ")";
+      if(m->getAccess() == clang::AS_public) llvm::outs() << "|public";
+      else if(m->getAccess() == clang::AS_private) llvm::outs() << "|private";
+      else if(m->getAccess() == clang::AS_protected) llvm::outs() << "|protected";
```

Comment:
Consider reducing code duplication for the code responsible for printing access specifiers